File: katas/string-calculator/python/solution/string_calculator.py
import re


def add(numbers: str) -> int:
    if numbers == "":
        return 0

    body, custom = _split_custom_delimiter(numbers)
    separators = f"[{re.escape(custom)}]" if custom else r"[,\n]"
    values = [int(part) for part in re.split(separators, body)]

    negatives = [v for v in values if v < 0]
    if negatives:
        raise ValueError("negatives not allowed: " + ",".join(str(v) for v in negatives))

    return sum(values)


def _split_custom_delimiter(input_str: str) -> tuple[str, str]:
    match = re.match(r"^//(.)\n(.*)$", input_str, flags=re.DOTALL)
    if not match:
        return input_str, ""
    return match.group(2), match.group(1)

File: katas/string-calculator/typescript/solution/stringCalculator.ts
export function add(numbers: string): number {
  if (numbers === "") return 0;

  const { body, delimiter } = splitCustomDelimiter(numbers);
  const parts = body === "" ? [] : body.split(new RegExp(`[,\n${escapeRegExp(delimiter)}]`));
  const values = parts.map(Number);

  const negatives = values.filter((v) => v < 0);
  if (negatives.length > 0) {
    throw new Error(`negatives not allowed: ${negatives.join(",")}`);
  }

  return values.reduce((a, b) => a + b, 0);
}

function splitCustomDelimiter(input: string): { body: string; delimiter: string } {
  const match = input.match(/^\/\/(.)\n(.*)$/s);
  if (!match) return { body: input, delimiter: "" };
  return { body: match[2], delimiter: match[1] };
}

function escapeRegExp(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}
